### PR TITLE
Pepper 575

### PR DESCRIPTION
**Pull request by denniscunningham**
## Context
PEPPER-575
Also see the discussion of the underlying issue possible approaches: [link](https://docs.google.com/document/d/1jyj-tQkryZsE9AMWtkVR4zL6-rJXFAcC35h86l7MyR8/edit?usp=sharing).

At a high level this was straightforward: map a DSS FIND_OUT answer into a DSM REF_SOURCE ID. However, the challenge was finding a way that was not too brittle: likely to break with a future change and be hard to detect. The changes here focus on detecting when things get out of sync.

Note: I really wanted to change the name RgpAutomaticProbandDataCreator to the simpler RgpProbandDataCreator, since the term 'Automatic' adds no value and is likely misleading. Since git history is lost with this I kept the name as is.

Also included in the PR are changes to get the Liquibase Maven plug-in configured correctly.

## Checklist

- [x ] I have labeled the type of changes involved using the `C-*` labels.
- [x ] I have assessed potential risks and labeled using the `R-*` labels.
- [x ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ x] I have considered security and privacy, and added `I-*` labels as needed
- [x ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ x] I have written automated positive tests
- [ x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [x ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 1]
```diff
@@ -37,8 +40,28 @@ protected boolean setDefaultData() {
...
             String participantId = StringUtils.isNotBlank(esProfile.getLegacyAltPid()) ? esProfile.getLegacyAltPid() : esProfile.getGuid();
             ParticipantData participantData = new ParticipantData(participantDataDao);
-            Optional<BookmarkDto> maybeFamilyIdOfBookmark = bookmarkDao.getBookmarkByInstance(RGP_FAMILY_ID);
-            Map<String, String> probandDataMap = extractProbandDefaultDataFromParticipantProfile(maybeFamilyIdOfBookmark);
+
+            Optional<BookmarkDto> familyIdOfBookmark = bookmarkDao.getBookmarkByInstance(RGP_FAMILY_ID);
+            if (familyIdOfBookmark.isEmpty()) {
+                // nothing in the call stack will handle a throw properly, so log the error and use the return value
+                // to unwind
+                logger.error("RGP family ID not found in Bookmark table");
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: Is error the right log level for this (WARN was what I was thinking) and would we benefit from logging additional information (participantId, requesting user, etc)?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 2]
```diff
@@ -1,9 +1,9 @@
 package org.broadinstitute.dsm.model.defaultvalues;
 
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: I _think_ we avoid wildcard imports where possible.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 3]
```diff
@@ -37,8 +40,28 @@ protected boolean setDefaultData() {
...
+            List<Activities> activities = elasticSearchParticipantDto.getActivities();
+            Map<String, String> probandDataMap =
+                    extractProbandDefaultData(activities, familyIdBookmarkDto.getValue());
+
+            try {
+                String refSourceId = convertReferralSources(getReferralSources(activities));
+                probandDataMap.put(REFERRAL_SOURCE_ID, refSourceId);
+            } catch (Exception e) {
+                // not good, but not fatal for this process
+                logger.error("Error deriving participant referral source for {}", participantId, e);
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: Silly question.  What shows up in the log statement related to e?  I only see one substitution in this log statement and if I'm reading that correctly, that's going to be the participantId.

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
+package org.broadinstitute.dsm.model.defaultvalues;
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java]: I remember you mentioning that you wanted to refactor the class name (RgpAutomaticProbandDataCreator) and ended up keeping it the same as the original.  Would you want to rename this class to the same name (RgpAutomaticProbandDataCreatorTest)?  Personally, I always like it when I search for a class by name in a codebase and the test shows up.

Comment (denniscunningham) [reply pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java]: Yup, I was aware of the discrepancy, but felt strongly about the incorrectness of the base class name. More strongly than I should, so will change this class name to match. :-|

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 2 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: Good spot. My guess is IntelliJ provided this but I missed it.

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+public class RgpProbandDataCreatorTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(RgpProbandDataCreatorTest.class);
+    private static List<Activities> activities;
+    private static String refSourceValues;
+    private static String refSourceDetails;
+    private static MockFieldSettingsDao mockDao;
+
+    @BeforeClass
+    public static void Setup() throws Exception {
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java]: Should be **setup**
We have a checkstyles configuration file that might not be loaded in your IDE that can help find these things.  See pepper-apis/checkstyle.xml and pepper-apis/docs/setting-up-checkstyle.md

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 6]
```diff
@@ -18,10 +18,13 @@
...
+import org.broadinstitute.dsm.util.proxy.jackson.ObjectMapperSingleton;
+
 
 public class RgpAutomaticProbandDataCreator extends BasicDefaultDataMaker {
 
     public static final String RGP_FAMILY_ID = "rgp_family_id";
+    public static final String REFERRAL_SOURCE_ID = "REF_SOURCE";
 
     @Override
     protected boolean setDefaultData() {
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java:44]: Is there a straightforward way to get this under unit test coverage?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 7]
```diff
@@ -97,6 +117,124 @@ private String getPhoneNumberFromActivities(List<Activities> activities) {
...
+        // get the REF_SOURCE IDs to verify the map is still in sync
+        List<Map<String, String>> refValues = ObjectMapperSingleton.readValue(
+                refPossibleValues, new TypeReference<List<Map<String, String>>>() {});
+
+        Set<String> refIDs = refValues.stream().map(m -> m.get("value")).collect(Collectors.toSet());
+        Set<String> refMapValues = new HashSet<>(refMap.values());
+        // "NA" and "MORE_THAN_ONE" are not mapped
+        if (refMapValues.size()+2 != refIDs.size() || !refIDs.containsAll(refMapValues)) {
+            // if IDs have diverged, don't stop this operation: log error and see if something is salvageable (below)
+            logger.error("RGP_MEDICAL_RECORDS_GROUP REF_SOURCE 'possibleValues' do not match REF_SOURCE map in " +
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: Log level appropriate?  Should it be WARN?
The question I'm thinking about in my head for these types of things: "Should we get an entry in the slack channel for this or is it somewhat expected and should we just log it?"

**Review by otchet-broad — COMMENTED** [review]
Overall - Great PR!  Thank you!
Logic seems straightforward to follow and the additional method javadoc style comments are a fantastic addition.

I've left some comment/questions in the code for your consideration.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 1 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: I think this is an error since we cannot proceed with the operation at this point. It replaces the former line 72 in this file, which currently throws but it only results in a warning. In any case,  I will restructure this to make more sense. And, of course, it should provide more useful details.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 7 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: I'll explain this better in the comment, but it is definitely something we should get alerted on. It's a runtime internal error, but there's not much recourse at this point (pubsub message handler) so I think it makes sense to press forward if we are able.

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java (hunk 5 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/defaultvalues/RgpProbandDataCreatorTest.java]: I also get a blizzard of checkstyle errors when running tests, so this did not stand out.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 6 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java:44]: Do you mean include the broader RGP data creator stuff, not just the new referral source stuff?

**Comment by denniscunningham** [comment]
Thx for your review!
I update the areas you flagged, and left a question about one the your comments.
Pls review again.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 6 above)
Comment (otchet-broad) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java:44]: More specifically, I'm looking to see if we have a test that covers the conditional at 45 being empty and the outcome of the catch at line 60.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java [hunk 8]
```diff
@@ -49,39 +76,72 @@ protected boolean setDefaultData() {
...
+            throw new RuntimeException(msg, e);
+        }
+
+        try {
+            String refSourceId = convertReferralSources(getReferralSources(activities));
+            probandDataMap.put(REFERRAL_SOURCE_ID, refSourceId);
+        } catch (Exception e) {
+            // not good: we could not convert referral source, but not fatal for this process.
+            // Use error level so humans get alerted to intervene and possibly fix the issue
+            logger.error("Error deriving participant referral source for participant {} and DDP instance {}: {}",
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java:122]: NIIIIIICE!

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/bookmark/BookmarkTest.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Test
+    public void getThenIncrementBookmarkValue() {
+        BookmarkDao bookmarkDao = mock(BookmarkDao.class);
+        String instanceId = "rgp_family_id";
+        long value = 1000;
+        int bookmarkId = 1;
+        BookmarkDto bookmarkDto = new BookmarkDto.Builder(value, instanceId).withBookmarkId(bookmarkId).build();
+
+        when(bookmarkDao.getBookmarkByInstance(instanceId)).thenReturn(Optional.of(bookmarkDto));
```
Comment (otchet-broad) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/bookmark/BookmarkTest.java:24]: Awesome!

**Review by otchet-broad — APPROVED** [review]

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 6 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java:44]: Addressed in latest commit.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java (hunk 3 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/defaultvalues/RgpAutomaticProbandDataCreator.java]: Good spot!. Fixed in latest commit.
